### PR TITLE
Echo and KVcounter integration benchmark tests

### PR DESCRIPTION
**Pull request by brooksmtownsend**
# Feature or Problem
This PR aims to add benchmark integration tests for the echo and kvcounter sample actors so that we can compare performance across new features and bug fixes in the OTP host.

# Release Information
`next` release

# Consumer Impact
The only impact this PR has is that it might slow down some of our integration tests by about a minute since they will be benchmarking.

# Testing
Tested on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

## Unit Test(s)
Modified echo, added echo_wasi benchmark

## Acceptance or Integration
Ideally, I'd like to include a way to upload benchmark results so we can compare over time.

## Manual Verification
I'm just running this locally:
```bash
MIX_ENV=test mix test test/host_core/benchmark/echo_test.exs
```

### CODE REVIEW

**Comment by connorsmith256** [comment]
> Ideally, I'd like to include a way to upload benchmark results so we can compare over time.

Would this be helpful? I.e. writing out benchmark results to a file and attaching it to a run as an artifact? https://github.com/actions/upload-artifact

File: host_core/test/host_core/benchmark/kvcounter_test.exs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      # Run the test at a few specified levels of parallelism, allowing for some warmup time to let compute calm down
+      parallel
+      |> Enum.each(fn parallel ->
+        IO.puts("Benchmarking with #{num_actors} actors and #{parallel} parallel requests")
+
+        Benchee.run(test_config,
+          warmup: 1,
+          time: 5,
+          parallel: parallel
+          # save: [path: "./echo_wasm32_wasi_#{num_actors}_#{parallel}.txt"] # TODO: this should be how we compare
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/kvcounter_test.exs:64]: There are a few places in this file that say "echo" which I think should say kvcounter

File: host_core/test/host_core/benchmark/kvcounter_test.exs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        @httpserver_key,
+        @httpserver_contract,
+        @httpserver_link,
+        "HttpServer.HandleRequest",
+        req
+      )
+
+    msg = %{
+      body: IO.iodata_to_binary(inv),
+      topic: "wasmbus.rpc.#{config.lattice_prefix}.#{@echo_wasi_key}",
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/kvcounter_test.exs:193]: I think this is the wrong topic?

File: host_core/test/host_core/benchmark/kvcounter_test.exs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    port = "8082"
+
+    # Make sure we don't log too much out of the HTTPserver provider
+    System.put_env("RUST_LOG", "info,warp=warn")
+
+    :ok = HostCoreTest.EventWatcher.wait_for_actor_start(evt_watcher, key)
+
+    # NOTE: Link definitions are put _before_ providers are started so that they receive
+    # the linkdef on startup. There is a race condition between provider starting and
+    # creating linkdef subscriptions that make this a desirable order for consistent tests.
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/kvcounter_test.exs:111]: Should lines 104-111 here also be in the echo test?

File: host_core/test/host_core/benchmark/echo_test.exs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+      |> Enum.each(fn parallel ->
+        IO.puts("Benchmarking with #{num_actors} actors and #{parallel} parallel requests")
+
+        Benchee.run(test_config,
+          warmup: 1,
+          time: 5,
+          parallel: parallel
+          # save: [path: "./echo_wasm32_unknown_#{num_actors}_#{parallel}.txt"] # TODO: this should be how we compare
+        )
+      end)
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/echo_test.exs:69]: Could some of this scaffolding be extracted and shared between the 3 tests? I'm thinking:
- setting the logger level at the beginning and end
- take a `test_config`, `num_actors`, and `parallel` and run the benchee stuff

**Review by connorsmith256 — COMMENTED** [review]
I only noticed this was still in draft when I went to submit the review, so ignore me if these comments aren't relevant :)

File: host_core/test/host_core/benchmark/echo_test.exs (hunk 4 above)
Comment (brooksmtownsend) [reply host_core/test/host_core/benchmark/echo_test.exs:69]: Yeah let me see if I can define some of this stuff in helpers

File: host_core/test/support/common.exs [hunk 5]
```diff
@@ -137,4 +137,46 @@ defmodule HostCoreTest.Common do
       policy_timeout_ms: 1_000
     }
   end
+
+  # Benchmarking common functions
```
Comment (connorsmith256) [inline host_core/test/support/common.exs:141]: Supernit: IMO these should stay in `benchmark/`. Regardless, these helpers are super useful

File: host_core/test/host_core/benchmark/echo_test.exs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    } do
+      on_exit(fn -> cleanup(pid, config) end)
+
+      # TODO: make configurable w/ benchee
+      num_actors = 1
+      parallel = [1, 10]
+
+      {msg, inv, port} = setup_echo_test(config, evt_watcher, @echo_key, @echo_path, num_actors)
+
+      # very noisy debug logs during bench
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/echo_test.exs:40]: Nit: I think these comments about noise can be removed now

File: host_core/test/host_core/benchmark/kvcounter_test.exs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        queryString: "",
+        method: "GET"
+      }
+      |> Msgpax.pack!()
+      |> IO.iodata_to_binary()
+
+    port = "8082"
+
+    # Make sure we don't log too much out of the HTTPserver provider
+    System.put_env("RUST_LOG", "info,warp=warn")
```
Comment (connorsmith256) [inline host_core/test/host_core/benchmark/kvcounter_test.exs:91]: Should this be included in the echo test too?

File: host_core/test/support/common.exs [hunk 8]
```diff
@@ -137,4 +137,46 @@ defmodule HostCoreTest.Common do
...
+  # Run a benchmark with specified config, repeating for each parallel argument if it's a list
+  def run_benchmark(test_config, num_actors, parallel \\ [1], warmup \\ 1, time \\ 5)
+      when is_list(parallel) do
+    parallel
+    |> Enum.each(fn p -> run_benchmark(test_config, num_actors, p, warmup, time) end)
+
+    :ok
+  end
+
+  def run_benchmark(test_config, num_actors, parallel, warmup, time)
```
Comment (connorsmith256) [inline host_core/test/support/common.exs:170]: Should `run_benchmark` call the pre/post functions so the caller doesn't have to explicitly adjust log levels? Or do you think there are cases we wouldn't want to set the level to info during a benchmark?

**Review by connorsmith256 — DISMISSED** [review]
Just nits. LGTM!

File: host_core/test/support/common.exs (hunk 8 above)
Comment (brooksmtownsend) [reply host_core/test/support/common.exs:170]: Actually, pretty fair. I'll make it call the functions and we can split that out if it's needed later 😄

File: host_core/test/host_core/benchmark/kvcounter_test.exs (hunk 7 above)
Comment (brooksmtownsend) [reply host_core/test/host_core/benchmark/kvcounter_test.exs:91]: it's there!

**Review by connorsmith256 — APPROVED** [review]

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**